pkg/loop/datasource.go

```diff
@@ -39,6 +41,13 @@ type dataSourceServer struct {
 }
 
 func (d *dataSourceServer) Observe(ctx context.Context, _ *emptypb.Empty) (*pb.ObserveReply, error) {
+	// Pipeline observations may return results after the context is cancelled, so we modify the
+	// deadline to give them time to return before the parent context deadline.
+	var cancel func()
+	ctx, cancel = utils.ContextWithDeadlineFn(ctx, func(old time.Time) time.Time {
+		return old.Add(-time.Until(old) / 10)
```

Comment:
I'm not certain that scaling by some % is necessarily the best solution here :thinking: 
In median's particular use case, we generally have several concurrent tasks fetching data from EAs, and then a final step calculate the answer. Ideally we want to maximize the time allocated to the EAs, and only reduce the deadline just enough to accommodate the final answer calculation, which is some fixed duration that is independent of the parent context deadline.

Reply:
could we introduce a prom histogram to measure how much extra time this takes and have it measure per job?

or is it easy to fix the root cause?

Reply:
I don't think we can measure it from the outside here. Something like that could be used inside the pipeline but would have to be a bit more general/dynamic IIUC.